File: day2/abramo_multithreading_acquisition/thread_example.py
# Multithreading example in Python
# Author: Jacopo Abramo, 26.09.2023

from napari.qt.threading import thread_worker
from napari.qt.threading import WorkerBase

MAX_COUNT = 100_000_000
NUM_THREADS = 8

@thread_worker
def long_count(N: int) -> None:
    print(f"Starting thread: Counting from {N} to 0...")
    while N > 0:
        N -= 1
    print(f"Done! Counter is {N}")

workers = [long_count(MAX_COUNT // NUM_THREADS) for _ in range(NUM_THREADS)]

for worker in workers:
    worker.start()

WorkerBase.await_workers()

File: day2/abramo_multithreading_acquisition/thread_example.c
/**
*  Multithreading example in C
*  This example can be compiled only on a Linux machine
*  or alternatively on Windows Subsystem Linux (WSL)
*  Author: Jacopo Abramo, 26.09.2023
*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <pthread.h>
#define NUM_THREADS 2
 
void *worker(void *maxCount)
{
    int count = *(int*)maxCount;
    int counter = 0;

    printf("Starting new thread (count = %d)...\n", count);
    while(count > 0)
    {
        count--;
    }

    printf("Thread terminated! Counter reached %d\n", count);
    return NULL;
}
 
int main(void)
{
    int maxCount = 100000000 / NUM_THREADS;

    pthread_t threads[NUM_THREADS];

    for(int i = 0; i < NUM_THREADS; i++)
    {
        pthread_create(&threads[i], NULL, worker, &maxCount);
    }

    for(int i = 0; i < NUM_THREADS; i++)
    {
        pthread_join(threads[i], NULL);
    }

    printf("Exiting from main program\n");
    return 0;
}